### ISSUE TITLE
Bithopper crash - greenlet?

### ISSUE THREAD

**Issue by Clipseza** [c00w/bitHopper#573]
Hi

I cant figure out why at random intervals bithopper simple gets killed , error log shows this related to it at last log entry. Any idea from this log what might cause it ? 

Traceback (most recent call last):
  File "C:\wamp\www\c00w-bitHopper-5468920\eventlet\greenpool.py", line 80, in _spawn_n_impl
    func(_args, *_kwargs)
  File "C:\wamp\www\c00w-bitHopper-5468920\eventlet\wsgi.py", line 514, in process_request
    proto = self.protocol(socket, address, self)
  File "C:\Python27\lib\SocketServer.py", line 641, in __init__
    self.finish()
  File "C:\wamp\www\c00w-bitHopper-5468920\eventlet\wsgi.py", line 456, in finish
    BaseHTTPServer.BaseHTTPRequestHandler.finish(self)
  File "C:\Python27\lib\SocketServer.py", line 694, in finish
    self.wfile.flush()
  File "C:\Python27\lib\socket.py", line 303, in flush
    self._sock.sendall(view[write_offset:write_offset+buffer_size])
  File "C:\wamp\www\c00w-bitHopper-5468920\eventlet\greenio.py", line 283, in sendall
    tail = self.send(data, flags)
  File "C:\wamp\www\c00w-bitHopper-5468920\eventlet\greenio.py", line 269, in send
    total_sent += fd.send(data[total_sent:], flags)
error: [Errno 10054] An existing connection was forcibly closed by the remote host

**Comment by c00w** [comment]
Your router is killing connections.

**Comment by Clipseza** [comment]
will contact the datacentre, i figured as much from the error message but also thought it could just be some generic error output not specificly pointing at datacentre killing it.

**Comment by Clipseza** [comment]
Just btw, I also get it to crash with same general error log except at the end

error: [errno 10053] Host machine closed connection etc etc

**Comment by c00w** [comment]
If they can give you how long sessions timeout we can lower the longpoll timeout so it is less than that.

**Comment by Clipseza** [comment]
Yeh let me find out

**Comment by Clipseza** [comment]
mm no connections being killed by datacenter is the official word from the network guru there. I still think that error is related to the machine its running on :/

**Comment by Clipseza** [comment]
Yeh the error that goes along with the one I posted displays connection was aborted by software on your host machine.

Im wondering if its not just because of too many open connections, Ive noticed it peaks sometimes before the connection reset around 1200 connection. Im running this on a windows 7 machine, can you tell me where in bithopper I can lower the open connection timeout cause I think its leaking no need to have that many and ever increasing connections open.

**Comment by c00w** [comment]
How many clients are you running through it and what miner are you using? You are probably hitting the connection limit on windows and it is killing sockets to make new ones. I'll take a look at bitHopper on my machine to see if its something on my end but it's probably the miner.

**Comment by Clipseza** [comment]
Right now between 20-30 workers but will grow as I am investing more funds the coming weeks into lower cost hashing monster. Ive tested it with every single miner from poclbm to cgminer and all in between, bithopper is running on a dedicated server so I dont know how a miner on a seperate machine could force bithopper on dedicated machine to kill connections/crash bithopper.

What I did notice looking at open connections on the bithopper server, it would remain around 250-350 open connections on a relative flatline for couple of minutes and then slowly but surely creep up and up and once it goes over 600(has hit as high as 1500) open connections it would randomly crash bithopper.

**Comment by Clipseza** [comment]
Btw if we can figure this out and resolve the crash I will send you a coin for dealing with my problems :P

**Comment by Clipseza** [comment]
Ye seems bithopper just leaks/opens way to many connections and then it gets killed, hopefully you can figure something out.

**Comment by Clipseza** [comment]
Ok one more observation, the ton of open connections seems to be related to excessive getwork(inefficient) spamming between moving pools, I noticed this by looking at bitcoinpool stats and my efficiency is poor as hell (alot more getworks than submits) 

Are you confident in the current getwork request setup, dont you think its getwork spamming to often?

**Comment by Clipseza** [comment]
Ah ok last attempt of giving some info.

Im thinking that perhaps a couple of pools is actually temp closing connections due to excessive getwork spam. Until now Ive only really ran 5-10gh over bithopper instance but I see no reason why I cant move all my miners over one bithopper , makes tracking much more convenient, Ive recently moved my whole 50gh onto bithopper and since then this is occuring.

Could you have a look at the getwork requests cause it does look like its asking for 2-3x as much getwork as it ever submits looking a bitcoinpool efficiency stats gives me that indication

**Comment by c00w** [comment]
Um, I'll take a look. All getwork requests we make are initiated by a client. It may be that your miner isn't waiting long enough  and cutting of requests before we can reply and making a new one. Switching to gevent should fix this.

**Comment by c00w** [comment]
Also we may be freezing in the middle of getworks causing a whole batch of them to time out because of sqlite and its stupid file io policy. I'm going to switch to berkely db which should solve some of this.

**Comment by c00w** [comment]
Turns out switching to berkelydb may not work. However there is a vacuum command in sqlite which should help. We probably should vacuum once per hour but I threw in a hot fix that cleans up on startup. It should solve problems for a bit.

**Comment by c00w** [comment]
And it now vacuums once per hour.

**Comment by c00w** [comment]
I'm going to switch to gevent also which should help with performance and just general error handling.

**Comment by Clipseza** [comment]
Hi

I am still getting occasional crash after connection been closed, I think those come from the pool that might see a ton of getworks over one account as botnet or inefficient miner or what not.

I wonder if it wouldnt be better to split traffic over multiple workers perhaps rather than all the traffic over one worker. For this to work bithopper would actually need to be able to do it, is that a huge amount of work ?

Some pools might have getwork limit per worker causing them to close connections on me which causes havoc shrug thats all i can think causing this.

**Comment by c00w** [comment]
Ok. Well I just updated the whole networking engine to use gevent. I also spun off the database into its own thread since it wasn't before due to excessive monkey patching.

I'll add multiple workers per pool to the feature list and once I get btcnet_info merged into bitHopper I'll redo the frontend websites and add multiple workers.

**Comment by Clipseza** [comment]
Great will have a look at latest update, I really think something like splitting work across multiple workers could come in handy.

Good stuff

**Comment by Clipseza** [comment]
Can you tell me where stats update now, seems stats.db doesnt update but webgui still shows stats counting o_0

**Comment by Clipseza** [comment]
Well not sure it seems you made error somewhere, latest updates killed bithopper getting this output:

Traceback (most recent call last):
  File "bitHopper.py", line 14, in <module>
    raise e
ImportError: cannot import name core
Traceback (most recent call last):
  File "bitHopper.py", line 14, in <module>
    raise e
ImportError: cannot import name core

**Comment by Clipseza** [comment]
I notice stats dont save to stats.db at all, erk! That happened ever since the vacuum code afaik. 

If you can have a look at your latest updates and see whats causing it to crash with error ontop and also let me know what I need to install extra with activepython if anything is needed.

**Comment by c00w** [comment]
You need to install gevent. pip install gevent should work on linux.

**Comment by Clipseza** [comment]
The database doesnt write tho, thats a big issue saving for logging my stats.

I got bithopper on windows box and pypm install gevent doesnt find the package

**Comment by c00w** [comment]
I removed the manual transactioning and its saving now.

**Comment by Clipseza** [comment]
great and gevent for pypm? Any idea about that?

**Comment by c00w** [comment]
You can grab gevent from http://www.lfd.uci.edu/~gohlke/pythonlibs/

**Comment by c00w** [comment]
Or http://code.google.com/p/gevent/downloads/detail?name=gevent-1.0a3.win32-py2.7.msi

**Comment by c00w** [comment]
Um, it looks like you can definitely get it for windows. I'm not sure how pypm works with installing packages externally.

**Comment by Clipseza** [comment]
I got gevent from external source, one last question.. How often does the stats write to the db file.

Is it possible to make it write directly(or with less delay, perhaps every 5seconds) without caching in memory?

**Comment by c00w** [comment]
It writes once every 60 seconds. We have a caching layer because sqlite can only do 12 sql transactions per minute unless you are on a ssd and it quickly dies in a fire without a caching layer. Now that it is properly split into another thread we can probably have it write once every 12 seconds safely.

**Comment by Clipseza** [comment]
Ye 12 sec caching would allready help

**Comment by c00w** [comment]
Okay. It's set at 20 seconds. I'll lower it if no errors appear.

**Comment by c00w** [comment]
I'm closing this issue and opening another one for lagger being broken on windows.
